### PR TITLE
SERVER-26442 Add perf test for project sort swap for sort pushdown

### PR DESCRIPTION
**Pull request by TedTuckman**

### CODE REVIEW

**Comment by TedTuckman** [comment]
I'm on the fence about the need to commit this -- it could be better just to run it with and without the patch to demonstrate there won't be a regression for this case. If so, I'll close the PR

File: src/workloads/execution/ProjectSort.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+- Name: InsertData
+  Type: Loader
+  Threads: 1
+  Phases:
+  - Repeat: 1
+    Database: &db test
+    Threads: 1
+    CollectionCount: 1
+    DocumentCount: 50000
+    BatchSize: &batchSize 100000
```
Comment (jimoleary) [inline src/workloads/execution/ProjectSort.yml:17]: Is the DocumentCount / batchSize mismatch deliberate? I assume it is given that it is reused later in the aggregations.

File: src/workloads/execution/ProjectSort.yml [hunk 2]
```diff
@@ -0,0 +1,121 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/query"
+Description: >
+  Check performance of a project followed by a sort that can be swapped in order to push down. Check the 
+  performance both without and without an index.
+
```
Comment (jimoleary) [inline src/workloads/execution/ProjectSort.yml:6]: If this is being committed:

-  can you list the important metrics to look at.
- add some keywords  to aid finding this workload (see [here](https://github.com/mongodb/genny/blob/fd3a0f5f26bff7e13037850595a16f07f697b442/src/workloads/scale/BigUpdate.yml#L9-L15) for an example).

There is a fair bit of duplication, is it worth introducing some [GlobalDefaults](https://github.com/mongodb/genny/blob/48b3fac2e0c8c29b5f196e1263acc8d89e3762ed/src/workloads/scale/InCacheSnapshotReads.yml#L11-L14) for (assuming you are committing it):

- Collection0
- the number of repeats
- the aggregation OperationCommand

Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:17]: Defaulted the collname, repeats, and the agg pipeline.

File: src/workloads/execution/ProjectSort.yml (hunk 1 above)
Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:28]: Fixed. I had the lower doc count for faster testing locally

File: src/workloads/execution/ProjectSort.yml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    Database: *db
+    Operations:
+    - OperationName: RunCommand
+      OperationCommand:
+        createIndexes: Collection0
+        indexes: &IndexSpec [{key: {"integer": 1}, name: "int_1"}]
+  - *Nop
+
+
+- Name: Quiesce
```
[the remark this answers is not in the document]
Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:67]: Done, put one after the index build. I thought it was only for writes? We don't care about the performance of the index build right?

[the remark this answers is not in the document]
Comment (davish) [reply src/workloads/execution/ProjectSort.yml:67]: I was under the impression that the Quiesce was useful to make sure the two phases didn't interfere with each other / overlap at all in a nondeterministic way that would add noise to the metrics. It's definitely cumbersome, but it doesn't hurt.

File: src/workloads/execution/ProjectSort.yml [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+Actors:
+- Name: InsertData
+  Type: Loader
+  Threads: 1
+  Phases:
+  - Repeat: 1
+    Database: &db test
+    Threads: 1
+    CollectionCount: 1
+    DocumentCount: 100000
```
Comment (davish) [inline src/workloads/execution/ProjectSort.yml:27]: You could use the same YAML variable for DocumentCount and BatchSize in order to make sure the mismatch doesn't happen if one of those numbers gets modified.

**Review by davish — APPROVED** [review]

File: src/workloads/execution/ProjectSort.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:71]: Done

File: src/workloads/execution/ProjectSort.yml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        cursor: {batchSize: *docCount}
+
+AutoRun:
+- When:
+    mongodb_setup:
+      $eq:
+      - replica
+      - replica-all-feature-flags
+      - shard-lite
+      - standalone
```
Comment (jimoleary) [inline src/workloads/execution/ProjectSort.yml:145]: Question: do we really need all these mongodb setups? Would standalone and standalone-classic-query-engine suffice?

File: src/workloads/execution/ProjectSort.yml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  - *Nop
+  - *Nop
+
+
+- Name: Quiesce
+  Type: RunCommand
+  Threads: 1
+  Phases:
+  - *Nop
+  - Repeat: 1
```
Comment (jimoleary) [inline src/workloads/execution/ProjectSort.yml:76]: Can you change this to use the QuiesceActor  (see either of the following examples: [1](https://github.com/mongodb/genny/blob/fd3a0f5f26bff7e13037850595a16f07f697b442/src/workloads/transactions/LLTAnalytics.yml#L236-L245) [2](https://github.com/mongodb/genny/blob/c38a522e44a12fec43558527f110998eaa817465/src/workloads/scale/TimeSeriesSortScale.yml#L41-L49)).

**Review by jimoleary — CHANGES_REQUESTED** [review]
nearly there, use the QuiesceActor and consider if we really need to run the workload on all the setups listed.

File: src/workloads/execution/ProjectSort.yml (hunk 5 above)
Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:128]: Done.

File: src/workloads/execution/ProjectSort.yml (hunk 6 above)
Comment (TedTuckman) [reply src/workloads/execution/ProjectSort.yml:77]: Done

File: src/workloads/execution/ProjectSort.yml (hunk 5 above)
Comment (jimoleary) [reply src/workloads/execution/ProjectSort.yml:128]: :+1:

**Review by jimoleary — APPROVED** [review]
LGTM mod the patch not being too noisy

**Review by davish — APPROVED** [review]

**Comment by jimoleary** [comment]
The results are not noisy (see the tables below) but there is also not a very large different between the Index and NoIndex results and the NoIndex version is actually marginally faster generally.

Is this what you expected? 

I looked in the log for [project_sort / classic](https://spruce.mongodb.com/task/sys_perf_linux_standalone_classic_query_engine_project_sort_patch_f387fa8cc75b9129fb1d57d302ed424349990d5e_62fa496c2fbabe58a4344776_22_08_15_13_27_22/logs?execution=0&logtype=task&sortBy=STATUS&sortDir=ASC) for [phase 6](https://evergreen.mongodb.com/task_log_raw/sys_perf_linux_standalone_classic_query_engine_project_sort_patch_f387fa8cc75b9129fb1d57d302ed424349990d5e_62fa496c2fbabe58a4344776_22_08_15_13_27_22/0?type=T#L1936) and the code seems to be doing a collscan:

```
{"t":{"$date":"2022-08-15T14:40:15.170+00:00"},"s":"I",  "c":"COMMAND",  "id":51803,   "ctx":"conn390","msg":"Slow query","attr":{"type":"command","ns":"test.Collection0","appName":"Genny","command":{"aggregate":"Collection0","pipeline":[{"$project":{"array":0}},{"$sort":{"integer":1}}],"cursor":{"batchSize":100000},"$db":"test","lsid":{"id":{"$uuid":"ac62cabb-bcd2-47de-b9e6-f4216fee1162"}}},"planSummary":"COLLSCAN","cursorid":3797792496619056244,"keysExamined":0,"docsExamined":100000,"hasSortStage":true,"numYields":104,"nreturned":100000,"queryHash":"5F5FC979","queryExecutionEngine":"classic","reslen":9088995,"locks":{"FeatureCompatibilityVersion":{"acquireCount":{"r":115}},"Global":{"acquireCount":{"r":115}}},"storage":{},"remote":"10.2.0.10:44792","protocol":"op_msg","durationMillis":307}}

```

-------------------------------

The sys_perf_linux_standalone_classic_query_engine_project_sort_patch_f387fa8cc75b9129fb1d57d302ed424349990d5e_62fa496c2fbabe58a4344776_22_08_15_13_27_22 look very stable.

|  test_name                          |  commit  |  metric                 |  exection  |  result                  |
|-------------------------------------|-------------|------------------------|---------------|------------------------|
|  AggregateIndex.AggregateIndex      |  2ed424  |  AverageLatency         |  0         |  323.0471238             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  AverageLatency         |  1         |  316.7287967             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  AverageLatency         |  2         |  322.330397              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  DurationTotal          |  0         |  2879                    |
|  AggregateIndex.AggregateIndex      |  2ed424  |  DurationTotal          |  1         |  2819                    |
|  AggregateIndex.AggregateIndex      |  2ed424  |  DurationTotal          |  2         |  2871                    |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency50thPercentile  |  0         |  319.4620785             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency50thPercentile  |  1         |  312.4628975             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency50thPercentile  |  2         |  318.4413445             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency80thPercentile  |  0         |  320.9364186             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency80thPercentile  |  1         |  315.35945119999997      |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency80thPercentile  |  2         |  320.9353906             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency90thPercentile  |  0         |  343.06032200000004      |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency90thPercentile  |  1         |  337.90158353333334      |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency90thPercentile  |  2         |  342.7571766666667       |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency95thPercentile  |  0         |  355.560381              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency95thPercentile  |  1         |  350.570868              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency95thPercentile  |  2         |  354.983079              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency99thPercentile  |  0         |  355.560381              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency99thPercentile  |  1         |  350.570868              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency99thPercentile  |  2         |  354.983079              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMax             |  0         |  355.560381              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMax             |  1         |  350.570868              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMax             |  2         |  354.983079              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMin             |  0         |  317.854297              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMin             |  1         |  311.78522               |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMin             |  2         |  317.58821               |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OperationThroughput    |  0         |  3.473428273706148e-06   |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OperationThroughput    |  1         |  3.5473572188719405e-06  |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OperationThroughput    |  2         |  3.4831069313827932e-06  |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OverheadTotal          |  0         |  8768.722667             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OverheadTotal          |  1         |  8741.582347             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OverheadTotal          |  2         |  8860.895722             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  AverageLatency         |  0         |  317.7203643             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  AverageLatency         |  1         |  314.2520424             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  AverageLatency         |  2         |  319.9082469             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  DurationTotal          |  0         |  2829                    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  DurationTotal          |  1         |  2797                    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  DurationTotal          |  2         |  2851                    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency50thPercentile  |  0         |  312.910502              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency50thPercentile  |  1         |  309.149729              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency50thPercentile  |  2         |  316.7774735             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency80thPercentile  |  0         |  320.1846256             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency80thPercentile  |  1         |  316.5785788             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency80thPercentile  |  2         |  320.56803440000004      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency90thPercentile  |  0         |  340.2809047666667       |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency90thPercentile  |  1         |  338.22102803333337      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency90thPercentile  |  2         |  340.97413750000004      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency95thPercentile  |  0         |  350.826743              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency95thPercentile  |  1         |  348.436053              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency95thPercentile  |  2         |  351.694809              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency99thPercentile  |  0         |  350.826743              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency99thPercentile  |  1         |  348.436053              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency99thPercentile  |  2         |  351.694809              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMax             |  0         |  350.826743              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMax             |  1         |  348.436053              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMax             |  2         |  351.694809              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMin             |  0         |  311.716766              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMin             |  1         |  308.004302              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMin             |  2         |  312.74024               |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OperationThroughput    |  0         |  3.5348179568752213e-06  |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OperationThroughput    |  1         |  3.575259206292456e-06   |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OperationThroughput    |  2         |  3.50754121360926e-06    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OverheadTotal          |  0         |  4087.395299             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OverheadTotal          |  1         |  4108.076754             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OverheadTotal          |  2         |  4169.509955             |


I ran another execution of sys_perf_linux_standalone_project_sort_patch_f387fa8cc75b9129fb1d57d302ed424349990d5e_62fa496c2fbabe58a4344776_22_08_15_13_27_22

|  test_name                          |  commit  |  metric                 |  exection  |  result                  |
|-------------------------------------|-------------|------------------------|---------------|------------------------|
|  AggregateIndex.AggregateIndex      |  2ed424  |  AverageLatency         |  0         |  321.0104776             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  AverageLatency         |  1         |  325.9624417             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  DurationTotal          |  0         |  2861                    |
|  AggregateIndex.AggregateIndex      |  2ed424  |  DurationTotal          |  1         |  2902                    |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency50thPercentile  |  0         |  317.308739              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency50thPercentile  |  1         |  322.0642725             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency80thPercentile  |  0         |  319.80450360000003      |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency80thPercentile  |  1         |  323.596952              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency90thPercentile  |  0         |  340.9024880333334       |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency90thPercentile  |  1         |  346.9018075000001       |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency95thPercentile  |  0         |  352.487981              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency95thPercentile  |  1         |  359.925285              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency99thPercentile  |  0         |  352.487981              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  Latency99thPercentile  |  1         |  359.925285              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMax             |  0         |  352.487981              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMax             |  1         |  359.925285              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMin             |  0         |  316.156535              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  LatencyMin             |  1         |  321.429324              |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OperationThroughput    |  0         |  3.4952813701502973e-06  |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OperationThroughput    |  1         |  3.4458993797381116e-06  |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OverheadTotal          |  0         |  8769.902297             |
|  AggregateIndex.AggregateIndex      |  2ed424  |  OverheadTotal          |  1         |  8758.021629             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  AverageLatency         |  0         |  316.950056              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  AverageLatency         |  1         |  320.6861172             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  DurationTotal          |  0         |  2826                    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  DurationTotal          |  1         |  2857                    |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency50thPercentile  |  0         |  313.5114565             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency50thPercentile  |  1         |  316.0668195             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency80thPercentile  |  0         |  318.17878360000003      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency80thPercentile  |  1         |  322.42485719999996      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency90thPercentile  |  0         |  335.96189630000003      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency90thPercentile  |  1         |  343.03391120000003      |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency95thPercentile  |  0         |  345.372767              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency95thPercentile  |  1         |  352.923954              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency99thPercentile  |  0         |  345.372767              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  Latency99thPercentile  |  1         |  352.923954              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMax             |  0         |  345.372767              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMax             |  1         |  352.923954              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMin             |  0         |  310.328766              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  LatencyMin             |  1         |  315.296863              |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OperationThroughput    |  0         |  3.5385704175513092e-06  |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OperationThroughput    |  1         |  3.5001750087504372e-06  |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OverheadTotal          |  0         |  4104.222918             |
|  AggregateNoIndex.AggregateNoIndex  |  2ed424  |  OverheadTotal          |  1         |  4072.342185             |

**Comment by TedTuckman** [comment]
We expect a collscan with the code as-is today, the ticket/change this is testing would change the index case to use the index. The original idea of the test was to catch if the non-index version got slower after the change, but it would also be nice to have the index numbers tracked. @jimoleary does that make sense?

**Comment by jimoleary** [comment]
That makes perfect sense, feel free to merge.

**Comment by TedTuckman** [comment]
evergreen merge

**Pull request merged**